File: merkle-tree/javascript/Transaction.js
const { sha256 } = require('./Utils')

class Transaction {
    constructor(id, amount, balance, timestamp) {
        this.id = id
        this.amount = amount
        this.balance = balance
        this.timestamp = timestamp

        this.hash = this.getHash()
    }

    getHash() {
        const { id, amount, balance, timestamp } = this
        const t = { id, amount, balance, timestamp }
        return sha256(JSON.stringify(t))
    }

    toJson() {
        const { id, amount, balance, timestamp } = this
        return { id, amount, balance, timestamp }
    }
}

module.exports = Transaction

File: merkle-tree/elixir/lib/transaction.ex
defmodule App.Transaction do
  alias App.Utils

  defstruct [:id, :amount, :balance, :transacted_at, :hash, :block_id]

  def init(data, block) do
    %__MODULE__{
      id: data.id,
      amount: data.amount,
      balance: data.balance,
      transacted_at: data.transacted_at
    }
    |> put_hash()
    |> put_block_address(block)
  end

  def hash(t) do
    t
    |> Map.take([:id, :amount, :balance, :transacted_at])
    |> Jason.encode!()
    |> Utils.sha256()
  end

  def put_hash(t) do
    t |> Map.put(:hash, hash(t))
  end

  def put_block_address(t, block) do
    t |> Map.put(:block_id, block.id)
  end
end
